### ISSUE TITLE
Cannot debug signal handler on windows

### ISSUE THREAD

**Issue by nd** [go-delve/delve#1745]
1. What version of Delve are you using (`dlv version`)?

```
Delve Debugger
Version: 1.3.2
Build: 5a947bceffa8aac96574f3f7f50cdede74fc2fa7
```

2. What version of Go are you using? (`go version`)?

```
go version go1.13.4 windows/amd64
```

3. What operating system and processor architecture are you using?

Windows/amd64

4. What did you do?

Debug the following program:

```
package main

import (
	"fmt"
	"os"
	"os/signal"
	"time"
)

func main() {
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, os.Interrupt)
	sig := <- sigs
	fmt.Printf("Got a signal '%v', start shutting down...\n", sig)
	time.Sleep(3 * time.Second) // break
	fmt.Println("Shutdown complete")
}

```

and send it CtrlC via GenerateConsoleCtrlEvent() WinApi call (https://github.com/kohsuke/winp/blob/master/native/sendctrlc/main.cpp).

5. What did you expect to see?

Program gets a signal and stops at breakpoint.

6. What did you see instead?

Program gets a signal (I can see the 'Got a signal 'interrupt', start shutting down...' in the output), but it doesn't stop at breakpoints after getting a signal.

Also it seems like adding a breakpoint makes the program terminate early: without a breakpoint I can see `Shutdown complete` in the output, with breakpoint - it is not there and program doesn't wait for 3 seconds.

Running delve with `--log-output rpc` doesn't show any errors. If there are other useful logging flags, please let me know, I will reproduce with them.

**Comment by aarzilli** [comment]
My guess is that what's happening is that delve ends up getting the signal and terminates, taking the inferior with it. The solution would involve telling windows to send the signal only to the inferior. On unix this has to do with properly configuring the process group and the foreground process. No clue how you would do that on windows.

**Comment by aarzilli** [comment]
out of curiosity if you go at cmd/dlv/cmds/commands.go:677 and add a print statement to the `case <-ch:` line, what happens?

**Comment by nd** [comment]
```
case sig := <-ch:
	fmt.Printf("Signal %v\n", sig)
```

prints `Signal interrupt` in both cases with and without a breakpoint in a signal handler code. Without a breakpoint program doesn't exit immediately and executes the signal handler code.

**Comment by aarzilli** [comment]
Comment out the signal.Notify line, if it works you can put a `if runtime.GOOS != "windows"` around it and make a PR.

**Comment by nd** [comment]
It seems like simply commenting out the Notify line doesn't help: in this case the program doesn't get a signal and exits without running a signal handler.

I've found that there is a dedicated [DBG_CONTROL_C](https://docs.microsoft.com/en-us/windows/win32/debug/writing-the-debugger-s-main-loop) exception code on windows. I've added a branch for it into proc_windows.go which sets `continueStatus = _DBG_EXCEPTION_NOT_HANDLED` and also sets some flag that next SIGINT should be ignored, and modified commands.go to check this flag. After that it is possible to debug a signal handler.

Let me explain my interest to signal handlers, maybe what I want is impossible or a bad idea.  People seems to use signals for graceful shutdown which release resources in their services and at the moment this doesn't work in debugger as program is killed with SIGKILL. My idea was to always finish debug session by sending the program SIGINT/Ctrl-C: if program doesn't have a signal handler it will exit, otherwise it will shutdown gracefully. This seems to work fine on linux.

What would you say? Is it strange to always end debug session this way? Maybe there is a better way to achieve it (e.g. https://github.com/go-delve/delve/issues/612)?

**Comment by aarzilli** [comment]
> I've found that there is a dedicated DBG_CONTROL_C exception code on windows. I've added a branch for it into proc_windows.go which sets continueStatus = _DBG_EXCEPTION_NOT_HANDLED

Sounds good.

> and also sets some flag that next SIGINT should be ignored, and modified commands.go to check this flag

Isn't this equivalent to always ignoring the interrupt signal when `runtime.GOOS == "windows"`? Also isn't there a risk that windows decides to send the ctrl-c to delve before sending it to the inferior? I think it's better to ignore it all the time on windows, it seems simpler and safer.

> Let me explain my interest to signal handlers, maybe what I want is impossible or a bad idea. People seem to use signals for graceful shutdown which releases resources in their services and at the moment this doesn't work in debugger as program is killed with SIGKILL. My idea was to always finish debug session by sending the program SIGINT/Ctrl-C: if program doesn't have a signal handler it will exit, otherwise it will shutdown gracefully. This seems to work fine on linux.

That's fine. I imagine the next problem you'll encounter is people that catch ctrl-c to do something other than terminating the program and then they can't terminate their program because you are sending SIGINT instead of killing it.

**Comment by nd** [comment]
It seems like without `signal.Notify(ch, syscall.SIGINT)` nor delve neither program get the signal. I've added some logging to `waitForDebugEvent` and without `Notify` it doesn't get an exception corresponding to Ctrl+C.

Calling `signal.Notify(ch, syscall.SIGINT)` but ignoring the signal works without any proc_windows.go modifications. Would that be a good PR?

> I imagine the next problem you'll encounter is people that catch ctrl-c to do something other than terminating the program and then they can't terminate their program because you are sending SIGINT instead of killing it.

True, it should be an option to use regular kill.

**Comment by aarzilli** [comment]
> It seems like without signal.Notify(ch, syscall.SIGINT) nor delve neither program get the signal. I've added some logging to waitForDebugEvent and without Notify it doesn't get an exception corresponding to Ctrl+C. Calling signal.Notify(ch, syscall.SIGINT) but ignoring the signal works without any proc_windows.go modifications.

Well, that's bizarre.

> Would that be a good PR?

Yes.

**Comment by nd** [comment]
> Well, that's bizarre.

I was wrong, delve gets a signal, but since it doesn't call `signal.Notify` its default handler is executed and delve exits.

**Comment by aarzilli** [comment]
Closing, fixed by 3c266810757cd4abafe46fc6c9203aeeef5e15bf
